sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathExpressionsSuite.scala

```diff
@@ -705,6 +818,42 @@ class MathExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper {
     checkEvaluation(BRound(-3.5, 0), -4.0)
     checkEvaluation(BRound(-0.35, 1), -0.4)
     checkEvaluation(BRound(-35, -1), -40)
+    checkEvaluation(checkDataTypeAndCast(
+      FloorExpressionBuilder.build(Seq(Literal(2.5), Literal(0)))), 2L)
+    checkEvaluation(checkDataTypeAndCast(
+      FloorExpressionBuilder.build(Seq(Literal(3.5), Literal(0)))), 3L)
+    checkEvaluation(checkDataTypeAndCast(
+      FloorExpressionBuilder.build(Seq(Literal(-2.5), Literal(0)))), -3L)
+    checkEvaluation(checkDataTypeAndCast(
+      FloorExpressionBuilder.build(Seq(Literal(-3.5), Literal(0)))), -4L)
+    checkEvaluation(checkDataTypeAndCast(
+      FloorExpressionBuilder.build(Seq(Literal(-0.35), Literal(1)))), -0.4)
+    checkEvaluation(checkDataTypeAndCast(
+      FloorExpressionBuilder.build(Seq(Literal(-35), Literal(-1)))), -40L)
+    checkEvaluation(checkDataTypeAndCast(
+      CeilExpressionBuilder.build(Seq(Literal(2.5), Literal(0)))), 3L)
+    checkEvaluation(checkDataTypeAndCast(
+      CeilExpressionBuilder.build(Seq(Literal(3.5), Literal(0)))), 4L)
+    checkEvaluation(checkDataTypeAndCast(
+      CeilExpressionBuilder.build(Seq(Literal(-2.5), Literal(0)))), -2L)
+    checkEvaluation(checkDataTypeAndCast(
+      CeilExpressionBuilder.build(Seq(Literal(-3.5), Literal(0)))), -3L)
+    checkEvaluation(checkDataTypeAndCast(
+      CeilExpressionBuilder.build(Seq(Literal(-0.35), Literal(1)))), -0.3)
+    checkEvaluation(checkDataTypeAndCast(
+      CeilExpressionBuilder.build(Seq(Literal(-35), Literal(-1)))), -30L)
+    checkEvaluation(checkDataTypeAndCast(
+      FloorExpressionBuilder.build(Seq(Literal(-0.1), Literal(0)))), -1L)
+    checkEvaluation(checkDataTypeAndCast(
+      FloorExpressionBuilder.build(Seq(Literal(5), Literal(0)))), 5L)
+    checkEvaluation(checkDataTypeAndCast(
+      FloorExpressionBuilder.build(Seq(Literal(3.1411), Literal(-3)))), 0L)
+    checkEvaluation(checkDataTypeAndCast(
+      CeilExpressionBuilder.build(Seq(Literal(-0.1), Literal(0)))), 0L)
+    checkEvaluation(checkDataTypeAndCast(
+      CeilExpressionBuilder.build(Seq(Literal(5), Literal(0)))), 5L)
+    checkEvaluation(checkDataTypeAndCast(
+      CeilExpressionBuilder.build(Seq(Literal(3.1411), Literal(-3)))), 1000L)
```

Comment:
I'm a bit concerned about this case. It means the result can be arbitrarily big, if the `scale` is a very large negative number.

Since we return long for negative scale, I think the scale can't be smaller than `-18`. However, this can be weird if the input is decimal, e.g. there is a column `col` of type `decimal(38, 0)`, it's weird if we fail `ceil(col, -30)`.

It seems to be that, `RoundCeil` should only accept decimal inputs if the scale is not 0. Then return type should consider both the input decimal type and the scale value.

Reply:
That's a good point. 

you wanted to say `RoundCeil should only accept decimal inputs if the scale is` **-ve** ? RoundCeil can accept decimal inputs if the scale is +ve, right ?

And this specific case don't apply to `RoundFloor` right ?

Reply:
`RoundFloor` has a similar issue, if the input is a negative value.

My proposal is, for `round/ceil(input, scale)` if `scale != 0`, the expected input type should be decimal type, and the type coercion framework should cast int, double, etc. to decimal type. The result type can be
```
case DecimalType.Fixed(p, s) =>
  if (scale < 0) {
    DecimalType(math.max(p - s, 1 - scala), 0)
  } else {
    DecimalType(p, math.min(s, scale))
  }
```

Reply:
@cloud-fan Do you think something like this would work ?

```
(child.dataType, scaleV.asInstanceOf[Int]) match {
    case (DecimalType.Fixed(p, s), x) if x < 0 =>
      val castedChild = Cast(child, DecimalType(math.max(p, 1 - x), 0))
      RoundCeil(castedChild, scale)
    case (DecimalType.Fixed(p, s), x) =>
      val castedChild = Cast(child, DecimalType(p, math.min(s, x)))
      RoundCeil(castedChild, scale)
    case _ => RoundCeil(child, scale)
}
```

Reply:
Something similar, but we shouldn't add cast on our own. We should rely on the framework
```
if (scaleV.asInstanceOf[Int] == 0) {
  Ceil(child)
} else {
  RoundCeil(child, scale)
}
...
case class RoundCeil ... {
  def inputTypes = Seq(DecimalType, IntegerType)
  def dataType = child.dataType match {
    case DecimalType.Fixed(p, s) =>
      if (scale < 0) {
        DecimalType(math.max(p, 1 - scale), 0)
      } else {
        DecimalType(p, math.min(s, scale))
      }
  }
}
```

Reply:
@cloud-fan So basically if the scale is `scale < 0`, this produces the result of `scale = 0` and the return type is always decimal. This don't shock me, but this will be different from the behaviour of the snow-flake `ceil`/`floor` function. Are you sure about this ?

REF: https://docs.snowflake.com/en/sql-reference/functions/ceil.html
`If scale_expr is negative, then it specifies the number of places before the decimal point to which to adjust the number. For example, if the scale is -2, then the result is a multiple of 100.`

Reply:
> this produces the result of scale = 0 and the return type is always decimal.

No, I'm not saying we should change the calculation. I mean we should return decimal type, so that it can be bigger than int64 (decimal type can hold 38 digits).

Reply:
@cloud-fan with `def inputTypes = Seq(DecimalType, IntegerType)`, the framework will cast all inputs to DecimalType and `-ve` scale for decimals are not allowed by default. If we put negative scale for decimal, we will receive the following error: 
`Negative scale is not allowed: -2. You can use spark.sql.legacy.allowNegativeScaleOfDecimal=true to enable legacy mode to allow it.`

In the code actuel, i think it would be enough to add `math.min(scale, -18)` for the `scale < 0` something as bellow.

(child.dataType, scaleV.asInstanceOf[Int]) match {
        case (_, 0) => Ceil(child)
        case (_, x) if x < 0 => RoundCeil(Cast(child, LongType), math.min(-18, scale))
        case (_: IntegralType, _) => RoundCeil(Cast(child, LongType), scale)
        case (_ : FloatType, _) => RoundCeil(Cast(child, DoubleType), scale)
        case _ => RoundCeil(child, scale)
      }

What do you think ?

Reply:
In my previous comment, I already proposed the way to deal with negative scale
```
case DecimalType.Fixed(p, s) =>
  if (scale < 0) {
    DecimalType(math.max(p, 1 - scale), 0)
  } else {
    DecimalType(p, math.min(s, scale))
  }
```

Reply:
@cloud-fan Here is the code snippet that evaluates DecimalType at `RoundBase`. Basically if we assign `s = 0` for the negative scale, then for `scale < 0`, it will produce the result of `scale = 0`.
```
def nullSafeEval(input1: Any): Any = {
    dataType match {
      case DecimalType.Fixed(_, s) =>
        val decimal = input1.asInstanceOf[Decimal]
        // Overflow cannot happen, so no need to control nullOnOverflow
        decimal.toPrecision(decimal.precision, s, mode)
...
}
```

I couldn't find any simpler way to deal negative scale with `def inputTypes = Seq(DecimalType, IntegerType)`. Do you have any other proposals to deal negative scale, please ? thanks.

Reply:
@cloud-fan do you have any update on this please ?

Reply:
It's easy to fix `round`, as it returns 0 if the negative scale is too big. e.g. `round(12.3, -3)` returns 0. But `ceil/floor` is more complicated as it may need to increase precision for large negative scale parameters.

My previous proposal is actually correct
```
def dataType ...
  case DecimalType.Fixed(p, s) =>
    if (scale < 0) {
      DecimalType(math.max(p, 1 - scale), 0)
    } else {
      DecimalType(p, math.min(s, scale))
    }
```
But we need to fix the eval function as well
```
def nullSafeEval(input1: Any): Any = {
    dataType match {
      case DecimalType.Fixed(p, s) =>
        val decimal = input1.asInstanceOf[Decimal]
        if (_scale >=0) {
          // Overflow cannot happen, so no need to control nullOnOverflow
          decimal.toPrecision(decimal.precision, s, mode)
        } else {
          Decimal(decimal.toBigDecimal.setScale(_scale, mode), p, s)
        }
...
}
```

Reply:
@cloud-fan Perfect, that does the work. Could you please confirm the `doGenCode` part below, please ?

```
case DecimalType.Fixed(_, s) =>
        s"""
          if (${_scale} >= 0) {
            ${ev.value} = ${ce.value}.toPrecision(${ce.value}.precision(), $s,
            Decimal.$modeStr(), true);
            ${ev.isNull} = ${ev.value} == null;
          } else {
            ${ev.value} = new java.math.BigDecimal(${ce.value}).
            setScale(${_scale}, java.math.BigDecimal.${modeStr});
            ${ev.isNull} = ${ev.value} == null;
          }"""
```

Reply:
The codegen can be smarter, because we know the result of `_scale >= 0` before generating code
```
if (_scale >= 0) {
  "java code"
} else {
  "java code"
}
```

Reply:
@cloud-fan Could you please give me the "java code" for the else part, please ? I'm getting compile exception on 'generated.java'.

Reply:
Can you push whatever you have first? Then I can edit the file directly with GitHub.

Reply:
@cloud-fan I pushed the code 

```
override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
    val ce = child.genCode(ctx)

    val evaluationCode = dataType match {
      case DecimalType.Fixed(p, s) =>
        if (_scale >= 0) {
          s"""
            ${ev.value} = ${ce.value}.toPrecision(${ce.value}.precision(), $s,
            Decimal.$modeStr(), true);
            ${ev.isNull} = ${ev.value} == null;"""
       } else {
          s"""
            ${ev.value} = ${ce.value}.toPrecision(${ce.value}.precision(), $s,
            Decimal.$modeStr(), true);
            ${ev.isNull} = ${ev.value} == null;"""
//          s"""
//          ${ev.value} = new java.math.BigDecimal(${ce.value}).
//            setScale(${_scale}, java.math.BigDecimal.${modeStr});"""
        }
```

Could you please check what's wrong with the following code for the else part, please ? thanks
```
s"""
          ${ev.value} = new java.math.BigDecimal(${ce.value}).
           setScale(${_scale}, java.math.BigDecimal.${modeStr});"""
```